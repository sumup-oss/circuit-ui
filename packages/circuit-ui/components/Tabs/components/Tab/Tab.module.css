.base {
  float: left;
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
  height: 100%;
  padding: var(--cui-spacings-kilo) var(--cui-spacings-tera);
  font-size: var(--cui-body-m-font-size);
  line-height: var(--cui-body-m-line-height);
  color: var(--cui-fg-subtle);
  white-space: nowrap;
  text-decoration: none;
  cursor: pointer;
  outline: none;
  background-color: var(--cui-bg-normal);
  border: none;
}

.base:hover {
  background-color: var(--cui-bg-normal-hovered);
}

.base:focus {
  outline: 0;
  box-shadow: inset 0 0 0 2px var(--cui-border-focus);
}

.base:focus::-moz-focus-inner {
  border: 0;
}

.base:focus:not(:focus-visible) {
  box-shadow: none;
}

.base:active {
  background-color: var(--cui-bg-normal-pressed);
}

.base[aria-selected="true"] {
  position: relative;
  font-weight: var(--cui-font-weight-semibold);
  color: var(--cui-fg-normal);
}

.base[aria-selected="true"]::after {
  position: absolute;
  right: 0;
  bottom: 0;
  left: 0;
  z-index: var(--cui-z-index-absolute);
  width: 100%;
  height: var(--cui-spacings-bit);
  content: " ";
  background: var(--cui-border-accent);
}
