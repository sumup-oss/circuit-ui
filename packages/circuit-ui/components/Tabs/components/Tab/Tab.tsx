/**
 * Copyright 2019, SumUp Ltd.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

'use client';

import { AnchorHTMLAttributes, ButtonHTMLAttributes, forwardRef } from 'react';

import { useComponents } from '../../../ComponentsContext/index.js';
import { EmotionAsPropType } from '../../../../types/prop-types.js';
import { clsx } from '../../../../styles/clsx.js';

import classes from './Tab.module.css';

export interface BaseProps {
  /**
   * Triggers selected styles of the component
   */
  selected?: boolean;
}

type LinkElProps = AnchorHTMLAttributes<HTMLAnchorElement>;
type ButtonElProps = ButtonHTMLAttributes<HTMLButtonElement>;

export type TabProps = BaseProps & LinkElProps & ButtonElProps;

const tabIndex = (selected: boolean) => (selected ? undefined : -1);

/**
 * Tab component that represents a single tab inside a Tabs wrapper
 */
export const Tab = forwardRef<HTMLButtonElement, TabProps>(
  ({ selected = false, className, ...props }, ref) => {
    const components = useComponents();
    const Link = components.Link as EmotionAsPropType;
    const Element = props.href ? Link : 'button';
    return (
      <Element
        ref={ref}
        role="tab"
        aria-selected={selected}
        tabIndex={tabIndex(selected)}
        className={clsx(classes.base, className)}
        {...props}
      />
    );
  },
);

Tab.displayName = 'Tab';
