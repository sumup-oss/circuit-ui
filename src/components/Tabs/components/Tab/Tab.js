/**
 * Copyright 2019, SumUp Ltd.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React from 'react';
import PropTypes from 'prop-types';
import styled from '@emotion/styled';
import { css } from '@emotion/core';

import { textMega, focusOutline } from '../../../../styles/style-helpers';

const defaultTabStyles = ({ theme }) => css`
  label: tab;
  ${textMega({ theme })};
  padding: ${theme.spacings.kilo} ${theme.spacings.tera};
  color: ${theme.colors.n700};
  text-decoration: none;
  cursor: pointer;
  background-color: ${theme.colors.white};
  border: none;
  white-space: nowrap;
  height: 100%;
  align-items: center;
  float: left;
  display: flex;
  flex-direction: column;
  justify-content: center;
  outline: none;

  &:hover {
    background-color: ${theme.colors.n100};
  }

  &:active {
    background-color: ${theme.colors.n200};
  }

  &:focus {
    ${focusOutline({ theme })};
  }
`;

const selectedTabStyles = ({ theme, selected }) =>
  selected &&
  css`
    label: tab--selected;
    position: relative;
    color: ${theme.colors.n900};

    &::after {
      content: ' ';
      position: absolute;
      bottom: 0;
      left: 0;
      right: 0;
      width: 100%;
      height: ${theme.spacings.bit};
      background: ${theme.colors.p500};
    }
  `;

const tabIndex = selected => (selected ? undefined : '-1');

/**
 * Tab component that represents a single tab inside a Tabs wrapper
 */
const StyledTab = styled('button')(defaultTabStyles, selectedTabStyles);

const Tab = ({ selected, ...props }) => (
  <StyledTab
    {...props}
    selected={selected}
    aria-selected={selected}
    tabIndex={tabIndex(selected)}
  />
);

Tab.propTypes = {
  /**
   * Triggers selected styles of the component
   */
  selected: PropTypes.bool
};

Tab.defaultProps = {
  selected: false,
  role: 'tab'
};

export { StyledTab };

/**
 * @component
 */
export default Tab;
