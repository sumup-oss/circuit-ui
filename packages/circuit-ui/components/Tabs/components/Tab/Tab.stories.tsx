/**
 * Copyright 2025, SumUp Ltd.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { Tab } from './Tab.js';

export default {
  title: 'Navigation/Tabs/Tab',
  component: Tab,
  tags: ['status:internal'],
  parameters: {
    chromatic: {
      disableSnapshot: true,
    },
  },
};

export const Base = () => (
  <div>
    <Tab>Button</Tab>
    <Tab href="#link">Link</Tab>
    <Tab selected>Selected</Tab>
  </div>
);
