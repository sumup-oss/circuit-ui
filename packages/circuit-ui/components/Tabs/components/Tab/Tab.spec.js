/**
 * Copyright 2019, SumUp Ltd.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React from 'react';

import Tab from './Tab';

describe('Tab', () => {
  describe('styles', () => {
    it('should render with default styles', () => {
      const actual = create(<Tab>content</Tab>);
      expect(actual).toMatchSnapshot();
    });

    it('should render with selected styles', () => {
      const actual = create(<Tab selected>content</Tab>);
      expect(actual).toMatchSnapshot();
    });
  });

  describe('business logic', () => {
    it('should accept a working ref', () => {
      const tref = React.createRef();
      const { container } = render(<Tab ref={tref} />);
      const button = container.querySelector('button');
      expect(tref.current).toBe(button);
    });
  });
});
