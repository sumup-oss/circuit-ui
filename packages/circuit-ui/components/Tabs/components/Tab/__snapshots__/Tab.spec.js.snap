// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tab styles should render with default styles 1`] = `
.circuit-0 {
  font-size: 1rem;
  line-height: 1.5rem;
  padding: 12px 32px;
  color: var(--cui-fg-subtle);
  -webkit-text-decoration: none;
  text-decoration: none;
  cursor: pointer;
  background-color: var(--cui-bg-normal);
  border: none;
  white-space: nowrap;
  height: 100%;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  float: left;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  outline: none;
}

.circuit-0:focus {
  outline: 0;
  box-shadow: 0 0 0 4px var(--cui-border-focus);
}

.circuit-0:focus::-moz-focus-inner {
  border: 0;
}

.circuit-0:focus:not(:focus-visible) {
  box-shadow: none;
}

.circuit-0:hover {
  background-color: var(--cui-bg-normal-hovered);
}

.circuit-0:active {
  background-color: var(--cui-bg-normal-pressed);
}

<button
  aria-selected="false"
  class="circuit-0 circuit-1"
  role="tab"
  tabindex="-1"
>
  content
</button>
`;

exports[`Tab styles should render with selected styles 1`] = `
.circuit-0 {
  font-size: 1rem;
  line-height: 1.5rem;
  padding: 12px 32px;
  color: var(--cui-fg-subtle);
  -webkit-text-decoration: none;
  text-decoration: none;
  cursor: pointer;
  background-color: var(--cui-bg-normal);
  border: none;
  white-space: nowrap;
  height: 100%;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  float: left;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  outline: none;
  position: relative;
  color: var(--cui-fg-normal);
}

.circuit-0:focus {
  outline: 0;
  box-shadow: 0 0 0 4px var(--cui-border-focus);
}

.circuit-0:focus::-moz-focus-inner {
  border: 0;
}

.circuit-0:focus:not(:focus-visible) {
  box-shadow: none;
}

.circuit-0:hover {
  background-color: var(--cui-bg-normal-hovered);
}

.circuit-0:active {
  background-color: var(--cui-bg-normal-pressed);
}

.circuit-0::after {
  content: ' ';
  position: absolute;
  bottom: 0;
  left: 0;
  right: 0;
  width: 100%;
  height: 4px;
  background: var(--cui-border-accent);
}

<button
  aria-selected="true"
  class="circuit-0 circuit-1"
  role="tab"
>
  content
</button>
`;
